File: src/cli/commands/validate.ts
import { defineCommand } from "citty";
import { z } from "zod";
import { resolveConfig } from "../../config/index.js";
import { validatePlan } from "../../schemas/validate.js";

const ValidateArgsSchema = z.object({
  slug: z.string().min(1, "slug is required"),
});

export const validateCommand = defineCommand({
  meta: {
    name: "validate",
    description: "Validate plan YAML files against schemas. Usage: plan-bender validate <slug>",
  },
  args: {
    slug: {
      type: "positional",
      description: "Plan slug to validate",
      required: true,
    },
  },
  async run({ args }) {
    const parsed = ValidateArgsSchema.safeParse(args);
    if (!parsed.success) {
      for (const issue of parsed.error.issues) console.error(issue.message);
      process.exit(1);
    }
    const config = resolveConfig(process.cwd());
    const result = validatePlan(args.slug, config);

    let hasOutput = false;

    if (result.prd.errors.length > 0) {
      console.error(`PRD (${result.prd.file}):`);
      for (const err of result.prd.errors)
        console.error(`  - ${err}`);
      hasOutput = true;
    }

    for (const issue of result.issues) {
      if (issue.errors.length > 0) {
        console.error(`Issue (${issue.file}):`);
        for (const err of issue.errors)
          console.error(`  - ${err}`);
        hasOutput = true;
      }
    }

    if (result.crossRef.length > 0) {
      console.error("Cross-reference errors:");
      for (const err of result.crossRef)
        console.error(`  - ${err}`);
      hasOutput = true;
    }

    if (result.cycles.length > 0) {
      console.error("Dependency cycles:");
      for (const err of result.cycles)
        console.error(`  - ${err}`);
      hasOutput = true;
    }

    if (result.valid) {
      console.log(`Plan "${args.slug}" is valid.`);
    } else {
      if (!hasOutput) console.error("Validation failed.");
      process.exit(1);
    }
  },
});

File: internal/cli/validate.go
package cli

import (
	"encoding/json"
	"fmt"

	"github.com/jasonraimondi/plan-bender/internal/config"
	"github.com/jasonraimondi/plan-bender/internal/planrepo"
	"github.com/jasonraimondi/plan-bender/internal/schema"
	"github.com/spf13/cobra"
)

// NewValidateCmd creates the validate command.
func NewValidateCmd() *cobra.Command {
	var jsonOutput bool

	cmd := &cobra.Command{
		Use:   "validate <slug>",
		Short: "Validate PRD and issue YAML files",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			slug := args[0]

			cfg, err := config.Load(".")
			if err != nil {
				return err
			}

			repo := planrepo.NewProd(cfg.PlansDir)
			result, err := repo.Validate(slug, cfg)
			if err != nil {
				return fmt.Errorf("opening plan %q: %w", slug, err)
			}

			if jsonOutput || isAgentMode(cmd) {
				return json.NewEncoder(cmd.OutOrStdout()).Encode(result)
			}

			printValidationResult(cmd, result)

			if !result.Valid {
				return fmt.Errorf("validation failed")
			}
			return nil
		},
	}

	cmd.Flags().BoolVar(&jsonOutput, "json", false, "output as JSON")
	return cmd
}

func printValidationResult(cmd *cobra.Command, result schema.PlanValidationResult) {
	out := cmd.OutOrStdout()

	if len(result.PRD.Errors) > 0 {
		fmt.Fprintf(out, "PRD (%s):\n", result.PRD.File)
		for _, e := range result.PRD.Errors {
			fmt.Fprintf(out, "  - %s\n", e)
		}
	}

	for _, ir := range result.Issues {
		if len(ir.Errors) > 0 {
			fmt.Fprintf(out, "Issue (%s):\n", ir.File)
			for _, e := range ir.Errors {
				fmt.Fprintf(out, "  - %s\n", e)
			}
		}
	}

	if len(result.CrossRef) > 0 {
		fmt.Fprintln(out, "Cross-reference errors:")
		for _, e := range result.CrossRef {
			fmt.Fprintf(out, "  - %s\n", e)
		}
	}

	if len(result.Cycles) > 0 {
		fmt.Fprintln(out, "Dependency cycles:")
		for _, e := range result.Cycles {
			fmt.Fprintf(out, "  - %s\n", e)
		}
	}

	if result.Valid {
		fmt.Fprintln(out, "Validation passed")
	}
}
